### ISSUE TITLE
[TEST] seqan3::test::tmp_filename interface and behaviour

### ISSUE THREAD

**Issue by smehringer** [seqan/product_backlog#281]
### Description 

The current behaviour has some quirks.

#### Current interface and behaviour:

```cpp
seqan3::test::tmp_filename const chopper_split_filename{"example.txt"};
// creates a seqan_test_XXXXXXXX directory (X replaced by random stuff)
// concatenates seqan_test_XXXXXXXX directory to "example.txt" via std::filesystem::path::operator/
// final path looks something like `/tmp/seqan_test_XXXXXXXX/example.txt` 
```

* **Quirk1**: The usage may be unintuitive because the name `tmp_filename` suggest that you already get a filename but you need to provide one nonetheless.

* **Quirk2**:
  ```cpp
  seqan3::test::tmp_filename const chopper_split_filename{"/example.txt"};
  // creates a seqan_test_XXXXXXXX directory (X replaced by random stuff)
  // concatenates seqan_test_XXXXXXXX directory to "example.txt" via std::filesystem::path::operator/ which replaces everything
  // final path looks like this `/example.txt` (probably not intended)
  ```

* **Quirk3**:
  ```cpp
  seqan3::test::tmp_filename const chopper_split_filename{""};
  // creates a seqan_test_XXXXXXXX directory (X replaced by random stuff)
  // appends empty path
  // final path looks like this `/tmp/seqan_test_XXXXXXXX` (and cannot be used as a filename)
  // so basically, this creates a temporary directory.
  ```

#### Possible alternative

```cpp
auto filename = seqan3::test::tmp_filename{}; // no other ctor provided
// creates a seqan_test_XXXXXXXX directory (X replaced by random stuff)
// appends a dummy filename.
// final path looks like this `/tmp/seqan_test_XXXXXXXX/test_file` 
// if possible, we can get rid of the subdirectory and create a temporary file  `/tmp/seqan_test_XXXXXXXX` instead of a directory

auto filename = seqan3::test::tmp_filename{}.replace_extension("txt"); // if extension matters
```

With additional tmp_directory feature:
```cpp
auto dir = seqan3::test::tmp_directory{}; // no other ctor provided
// creates a seqan_test_XXXXXXXX directory (X replaced by random stuff)

std::filesystem::path p_in_dir = dir / "foo.txt"; // p_in_dir == "/tmp/seqan_test_XXXXXXXX/foo.txt"
// our operator/ will not behave exactly like the stl one, namely it will NOT replace the whole string if name starts with '/'
std::filesystem::path p_in_dir2 = dir / "/foo.txt"; // p_in_dir2 == "/tmp/seqan_test_XXXXXXXX/foo.txt"

// auto filename = seqan3::test::tmp_directory{}.get_path() ; // explicitly forbidden
```

#### General behaviour

* `seqan3::test::tmp_filename` owns the file and the directory and therefore must delete the file after lifetime ends.
* `seqan3::test::tmp_directory` owns the directory and therefore must delete the file after lifetime ends.

**Comment by marehr** [comment]
2021-01-25 Core-Meeting:

* For purist reasons we only want `seqan3::test::tmp_directory` with the following features:
  * Creates unique folder within tmp-Path
  * On destruction: Check whether tmp directory is empty; if not warn about left-over files; and delete all left-over files
    * warn with std::cerr about deletion
    * Test could add `EXPECT_TRUE((tmp_directory.is_empty()));` 
  * seqan3::test::tmp_directory{} / "foo" / "/hello"
    * seqan3::test::tmp_directory{} / "foo" returns seqan3::test::tmp_directory::path which inherits std::filesystem
    * seqan3::test::tmp_directory{} / "foo" / "../" should not be allowed
    * seqan3::test::tmp_directory{} / "foo" "/hello" should not be allowed
    * general idea is if target directory leaves our generated tmp directory it should throw
    * also for /=

**Comment by smehringer** [comment]
### Core Meeting 13.10.2022

* Remove the fucntion `tmp_directory::clean()`
* The destructor will automatically clean everything without warning

**Comment by eseiler** [comment]
Resolved by [#3063](https://github.com/seqan/seqan3/pull/3062) and [#3063](https://github.com/seqan/seqan3/pull/3063)
